src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,243 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                }
+                else
                 {
-                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    if (numberScale < 0)
                     {
                         result = default;
                         return false;
+
                     }
-                }
+                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
 
-                if (partialDigitCount > 0)
-                {
-                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+                    Span<uint> buffer = new uint[bufferSize];
+
+                    int bufferInd = buffer.Length - 1;
+                    uint currentBlock = 0;
+                    int shiftUntil = (totalDigitCount - 1) % MaxPartialDigits;
+                    int remainingIntDigitCount = totalDigitCount;
+                    foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                    {
+                        var digitsChunkSpan = digitsChunk.Span;
+                        ReadOnlySpan<char> intDigitsSpan = digitsChunkSpan.Slice(0, Math.Min(remainingIntDigitCount, digitsChunkSpan.Length));
+
+                        for (int i = 0; i < intDigitsSpan.Length; i++)
+                        {
+                            char digitChar = intDigitsSpan[i];
+                            Debug.Assert(char.IsDigit(digitChar));
+                            currentBlock = currentBlock * 10 + digitChar - '0';
+                            if (shiftUntil == 0)
+                            {
+                                buffer[bufferInd] = currentBlock;
+                                currentBlock = 0;
+                                bufferInd--;
+                                shiftUntil = MaxPartialDigits;
+                            }
+                            shiftUntil--;
+                        }
+                        remainingIntDigitCount -= intDigitsSpan.Length;
+
+                        ReadOnlySpan<char> fracDigitsSpan = digitsChunkSpan.Slice(intDigitsSpan.Length);
+                        for (int i = 0; i < fracDigitsSpan.Length; i++)
+                        {
+                            char digitChar = fracDigitsSpan[i];
+                            if (digitChar == '\0')
+                            {
+                                break;
+                            }
+                            if (digitChar != '0')
+                            {
+                                result = default;
+                                return false;
+                            }
+                        }
+                    }
+                    Debug.Assert(bufferInd == -1);
+
+                    unsafe
+                    {
+                        Span<uint> newBuffer = new uint[bufferSize];
```

Comment:
Given we end up throwing this buffer away on L765 (https://github.com/dotnet/runtime/pull/55121/files#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R765), we might be able to just rent this one as well?

Might not be worth it, however.

Reply:
It's hard to tell because of the deep nesting, but `buffer` and `newBuffer` are allocated only once. After [L734](https://github.com/dotnet/runtime/pull/55121/files/7d89c46821a9c476611419116fc8e1e0e885635a#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R734), which is the place that this buffer is thrown away, there is no place that can reuse this buffer as far as I know.

Reply:
Right, that's why I was suggesting we simply rent this buffer as well. Not to allow reuse but to avoid having an allocation that only lives for the length of this method.

Reply:
I see. I don't know it will improve the performance or not, so I'll take a benchmark and compare it.